src/V3Udp.cpp

```diff
@@ -0,0 +1,235 @@
+                if (nodep->isInput()) {
+                    m_inputVars.push_back(nodep);
+                } else {
+                    m_outputVars.push_back(nodep);
+                }
+                if ((m_inputVars.size() == 0) && (m_outputVars.size() == 1)) {
+                    m_isFirstOutput = true;
+                }
+            }
+        }
+        iterateChildren(nodep);
+    }
+    void visit(AstUdpTable* nodep) override {
+        FileLine* const fl = nodep->fileline();
+        m_inputNum = m_inputVars.size();
+        m_outputNum = m_outputVars.size();
+        if (m_outputNum != 1) {
+            m_outputVars.back()->v3error(
+                m_outputNum << " output ports for udp table, there must be one output port");
+        }
+        if (!m_isFirstOutput && m_outputNum) {
+            m_inputVars[0]->v3error("The first port must be the output port");
+        }
+        m_oFieldVarp = m_outputVars[0];
+
+        m_alwaysBlockp = new AstAlways{fl, VAlwaysKwd::ALWAYS, nullptr, nullptr};
+        fl->warnOff(V3ErrorCode::LATCH, true);
+        iterateChildren(nodep);
+
+        nodep->replaceWith(m_alwaysBlockp);
+    }
+    void visit(AstUdpTableLine* nodep) override {
+        FileLine* const fl = nodep->fileline();
+        if (!nodep->udpIsCombo() && !m_oFieldVarp->isBitLogic()) {
+            m_oFieldVarp->v3error("For sequential udp, the output var should be the reg type");
+        }
+        if (nodep->udpIsCombo() && m_oFieldVarp->isBitLogic()) {
+            m_oFieldVarp->v3error(
+                "For combinational udp, the output var shouldn't be the reg type");
+        }
+        AstNode* iNodep = nodep->iFieldp();
+        AstNode* oNodep = nodep->oFieldp();
+        uint32_t inputvars = 0;
+        AstSenTree* edgetrigp = nullptr;
+
+        AstLogAnd* logandp = new AstLogAnd(fl, new AstConst{fl, AstConst::BitTrue{}},
+                                           new AstConst{fl, AstConst::BitTrue{}});
+
+        for (AstVar* itr : m_inputVars) {
+            if (!iNodep) break;
+            inputvars++;
+            if (AstUdpTableLineVal* linevalp = VN_CAST(iNodep, UdpTableLineVal)) {
+                string valName = linevalp->name();
+                AstVarRef* const referencep = new AstVarRef{fl, itr, VAccess::READ};
+                if (isEdgeTrig(valName)) {
+                    if (nodep->udpIsCombo()) {
+                        linevalp->v3error(
+                            "There should not be a edge trigger for combinational UDP table line");
+                    }
+                    if (edgetrigp) {
+                        linevalp->v3error("There can be only one edge tigger signal");
+                    }
+                    edgetrigp = new AstSenTree{
+                        fl, new AstSenItem{fl, VEdgeType::ET_BOTHEDGE,
+                                           new AstVarRef{fl, itr, VAccess::READ}}};
+                }
+                if (valName == "0" || valName == "f")
```

Comment:
Do we convert to lower case before this?  Perhaps verilog.l should just upcase to avoid later in many places.

Please make sure upper case tested.

Reply:
We convert it in the `isEdgeTrig` function. Added upper case descriptors to the `t_udp_sequential.v` test